File: lib/services/platform/clock.dart
/// Wall-clock time, injected everywhere a service needs "now" so tests
/// can control it (architecture.md §18.1, §18.4).
abstract interface class Clock {
  /// The current time.
  ///
  /// Never throws.
  DateTime now();
}

/// A [Clock] backed by the real system clock.
final class SystemClock implements Clock {
  /// Creates a clock backed by [DateTime.now].
  const new();

  @override
  DateTime now() => DateTime.now();
}

File: backend/app/services/clock.py
"""The one source of the current time, so tests can pass their own.

Mirrors the Flutter side's `services/platform/clock.dart`: functions that care
about freshness take the time as a parameter rather than reading the wall clock
themselves, which is what makes cache expiry testable without sleeping.

Times are naive UTC throughout. SQLite does not preserve a timezone, so storing
aware datetimes would read back naive and compare wrongly; keeping one naive
UTC convention everywhere removes that whole class of bug.
"""

from datetime import UTC, datetime


def utc_now() -> datetime:
    """Returns the current UTC time, without a timezone attached."""
    return datetime.now(tz=UTC).replace(tzinfo=None)
